src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/MetadataUtilsSmi.cs

```diff
@@ -0,0 +1,625 @@
+            InvalidSqlDbType,               // null reference doesn't have an inferable SqlDbType
+            SqlDbType.SmallInt,             // System.Int16
+            SqlDbType.Int,                  // System.Int32
+            SqlDbType.BigInt,               // System.Int64
+            InvalidSqlDbType,               // System.SByte doesn't have an inferable SqlDbType
+            SqlDbType.Real,                 // System.Single
+            SqlDbType.NVarChar,             // System.String
+            InvalidSqlDbType,               // System.UInt16 doesn't have an inferable SqlDbType
+            InvalidSqlDbType,               // System.UInt32 doesn't have an inferable SqlDbType
+            InvalidSqlDbType,               // System.UInt64 doesn't have an inferable SqlDbType
+            InvalidSqlDbType,               // System.Object doesn't have an inferable SqlDbType
+            SqlDbType.VarBinary,            // System.ByteArray
+            SqlDbType.NVarChar,             // System.CharArray
+            SqlDbType.UniqueIdentifier,     // System.Guid
+            SqlDbType.VarBinary,            // System.Data.SqlTypes.SqlBinary
+            SqlDbType.Bit,                  // System.Data.SqlTypes.SqlBoolean
+            SqlDbType.TinyInt,              // System.Data.SqlTypes.SqlByte
+            SqlDbType.DateTime,             // System.Data.SqlTypes.SqlDateTime
+            SqlDbType.Float,                // System.Data.SqlTypes.SqlDouble
+            SqlDbType.UniqueIdentifier,     // System.Data.SqlTypes.SqlGuid
+            SqlDbType.SmallInt,             // System.Data.SqlTypes.SqlInt16
+            SqlDbType.Int,                  // System.Data.SqlTypes.SqlInt32
+            SqlDbType.BigInt,               // System.Data.SqlTypes.SqlInt64
+            SqlDbType.Money,                // System.Data.SqlTypes.SqlMoney
+            SqlDbType.Decimal,              // System.Data.SqlTypes.SqlDecimal
+            SqlDbType.Real,                 // System.Data.SqlTypes.SqlSingle
+            SqlDbType.NVarChar,             // System.Data.SqlTypes.SqlString
+            SqlDbType.NVarChar,             // System.Data.SqlTypes.SqlChars
+            SqlDbType.VarBinary,            // System.Data.SqlTypes.SqlBytes
+            SqlDbType.Xml,                  // System.Data.SqlTypes.SqlXml
+            SqlDbType.Structured,           // System.Collections.IEnumerable, used for TVPs it must return IDataRecord
+            SqlDbType.Structured,           // System.Collections.Generic.IEnumerable<Microsoft.SqlServer.Server.SqlDataRecord>
+            SqlDbType.Time,                 // System.TimeSpan
+            SqlDbType.DateTimeOffset,       // System.DateTimeOffset
+        };
+
+        // Hash table to map from clr type object to ExtendedClrTypeCodeMap enum
+        // this HashTable should only be accessed from DetermineExtendedTypeCode and class ctor for setup.
+        private static readonly Hashtable s_typeToExtendedTypeCodeMap;
```

Comment:
Any reason to use ```Hashtable``` here instead of ```Dictionary<Type, ExtendedClrTypeCode>```?

Looks like you could then also avoid the explicit static cctor by doing:
```C#
private static readonly Dictionary<Type,ExtendedClrTypeCode> s_typetoExtendedTypeCodeMap = 
    new Dictionary<Type,ExtendedClrTypeCode>
    {
        { typeof(System.Boolean), ExtendedClrTypeCode.Boolean },
        { typeof(System.Byte), ExtendedClrTypeCode.Byte },
        ...
        typeof(System.DateTimeOffset), ExtendedClrTypeCode.DateTimeOffset }
    };
```